cng/rsa.go

```diff
@@ -232,18 +232,36 @@ func VerifyRSAPSS(pub *PublicKeyRSA, h crypto.Hash, hashed, sig []byte, saltLen
 	return keyVerify(pub.hkey, unsafe.Pointer(&info), hashed, sig, bcrypt.PAD_PSS)
 }
 
-func SignRSAPKCS1v15(priv *PrivateKeyRSA, h crypto.Hash, hashed []byte) ([]byte, error) {
+// SignRSAPKCS1v15 calculates the signature of hashed using
+// RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5.  Note that hashed must
+// be the result of hashing the input message using the given hash
+// function. If hash is zero, hashed is signed directly.
+func SignRSAPKCS1v15(priv *PrivateKeyRSA, hash crypto.Hash, hashed []byte) ([]byte, error) {
 	defer runtime.KeepAlive(priv)
-	info, err := newPKCS1_PADDING_INFO(h)
+	if hash != crypto.Hash(0) {
+		if len(hashed) != hash.Size() {
+			return nil, errors.New("crypto/rsa: input must be hashed message")
+		}
+	}
+	info, err := newPKCS1_PADDING_INFO(hash)
 	if err != nil {
 		return nil, err
 	}
 	return keySign(priv.hkey, unsafe.Pointer(&info), hashed, bcrypt.PAD_PKCS1)
 }
 
-func VerifyRSAPKCS1v15(pub *PublicKeyRSA, h crypto.Hash, hashed, sig []byte) error {
+// VerifyPKCS1v15 verifies an RSA PKCS #1 v1.5 signature.
+// hashed is the result of hashing the input message using the given hash
+// function and sig is the signature. A valid signature is indicated by
+// returning a nil error. If hash is zero then hashed is used directly.
+func VerifyRSAPKCS1v15(pub *PublicKeyRSA, hash crypto.Hash, hashed, sig []byte) error {
 	defer runtime.KeepAlive(pub)
-	info, err := newPKCS1_PADDING_INFO(h)
+	if hash != crypto.Hash(0) {
+		if len(hashed) != hash.Size() {
+			return errors.New("crypto/rsa: input must be hashed message")
```

Comment:
Same in here, maybe we can create a custom error type for this like
```go
type ErrInvalidHashLength struct {
	Expected int
	Actual   int
}

func (e *ErrInvalidHashLength) Error() string {
	return "crypto/rsa: invalid hash length, expected " + strconv.Itoa(e.Expected) + ", got " + strconv.Itoa(e.Actual)
}
```

Reply:
I would agree if you say we programmically not checking the returned actual hash size and handling it, lets at least add it to error message, so we can see at least if what size is returned hash length, so maybe it can give us some clue about debugging

Reply:
FWIW, I trust upstream to have checked whether or not their error message strings have the potential to reveal sensitive information if the err message is exposed due to bad app code. I think that's another reason to stick with upstream's strings in the general case--we have the risk of making a bad situation even worse if we add more info. (We can change the tests, after all, but this is another reason not to, beyond maintenance cost. 😄)